#!/bin/bash
#PBS -N pseudo-haploid
#PBS -S /bin/bash
#PBS -l mem=15gb
#PBS -l walltime=08:00:00 
#PBS -l nodes=1:ppn=2

module load gcc/6.2.0
source activate /scratch/cliu7/miniconda_env/bioinformatics/

CHR=$1

call_pseudo_haploid() {
    project='/home/cliu7/ancient_ACA'
    scratch='/scratch/cliu7/ancient_ACA'
    ref_genome='/gpfs/data/novembre-lab/reference_genome/hs37d5.fa'
    param="${project}/data/params/pseudo_haploid"
    geno_dir="${scratch}/data/intermediate"

    snp_prefix=${param}/$1
    bam_prefix=${param}/$2
    sample_id=${param}/$3
    final_ind_prefix=${param}/$4
    out=${geno_dir}/$5

    # produce snp file
    awk -vOFS=$'\t' '{print $2,$4}' ${snp_prefix}.snp > ${snp_prefix}.pos

    # pileup and random draw representative haploid
    samtools mpileup -RB -q30 -Q30 \
	     -b ${bam_prefix}.txt \
	     -l ${snp_prefix}.pos \
	     -f ${ref_genome} | \
	pileupCaller --sampleNameFile ${sample_id}.txt \
		     --samplePopName POP \
		     -f ${snp_prefix}.snp \
		     -o EigenStrat \
		     -e ${out}

    # rename and clean up
    for i in geno snp ind
    do
	mv ${out}.${i}.txt ${out}.${i}
    done

    cp ${final_ind_prefix}.txt  ${out}.ind

}

call_pseudo_haploid chb_ha_pos/CHB_HA.chr${CHR}.c1b aACA_wgs_bam aACA_wgs_sample aACA_wgs_ind aACA_wgs.CHR${CHR}
#call_pseudo_haploid chb_ha_pos/CHB_HA.chr${CHR}.c1b aACA_wgsread_bam aACA_wgsread_sample aACA_wgsread_ind aACA_wgsread.CHR${CHR}
